### PR TITLE
Add optimization summary documents

### PR DESCRIPTION
**Pull request by a7ehuo**
Added the following four documents on the summary of the JIT optimizations. These four documents are extracted
from the document written by Vijay - `"Summary of optimizations in the Testarossa JIT"` stored in Notes DB `InsideTRJIT`.

- LocalOptimizationsSummary.md
- GlobalOptimizationsSummary.md
- ControlFlowOptimizationsSummary.md
- LoopOptimizationSummary.md

Authored-By: Vijay Sundaresan <vijaysun@ca.ibm.com>

Signed-off-by: Annabelle Huo <Annabelle.Huo@ibm.com>

### CODE REVIEW

**Comment by a7ehuo** [comment]
@vijaysun-omr @hzongaro May I ask you to review these documents? Thank you!
@0xdaryl fyi

**Review by hzongaro — COMMENTED** [review]
Thanks for pulling this together, Annabelle!  I haven't finished reviewing yet, but I thought I would submit some of my comments for the first parts.

**Comment by hzongaro** [comment]
The commit comment mentions a database that's not accessible to most people in the community.  I would suggest that you mention that to make that clear.

File: doc/compiler/optimizer/GlobalOptimizationsSummary.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+fields the presence of a GC point before any user-specified definition means that
+the field must be zero initialized. Peeking inside calls is done to maximize the
+number of explicit field initializations seen by the analysis. In addition, for
+some special array allocations (e.g. char arrays that are used by `String`, often do
+not need zero initialization as they are filled in by `arraycopy` calls immediately),
+we can skip zero initialization completely.
+
+## Global Copy Propagation (Dead Store Elimination, Dead Structure Removal)
+These three analyses use use/def info to perform the traditional optimizations. The
+analysis is only done for locals (statics/fields omitted). Dead structure removal
```
Comment (dsouzai) [inline doc/compiler/optimizer/GlobalOptimizationsSummary.md:79]:
```suggestion
analysis is only done for locals (statics/fields are omitted). Dead structure removal
```

File: doc/compiler/optimizer/GlobalOptimizationsSummary.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+basically tries to find a region of code with no side-effects and having no definition
+that is used outside the structure; in such cases the (dead) region of code can be
+eliminated.
+
+## Redundant Monitor Elimination
+Redundant Monitor Elimination uses value numbering to perform nested monitor removal
+(e.g. syncenter on the same object twice followed by two syncexits). Use value
```
Comment (dsouzai) [inline doc/compiler/optimizer/GlobalOptimizationsSummary.md:86]:
```suggestion
(e.g. syncenter on the same object twice followed by two syncexits). It uses value
```

File: doc/compiler/optimizer/GlobalOptimizationsSummary.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+GRA is a tactical solution for achieving the effect of global register assignment,
+even though the underlying code generators have local register allocators. Basic idea
+is that global register dependencies are introduced in the IL trees; these dependencies
+(when evaluated by the code generators) ensure that a certain value is in a certain
+register on all control flow paths into a basic block. Stores/loads from memory are
+changed into stores/loads from registers in the IL (register numbers specify which
+physical register the value is placed into). Global register candidates are chosen
+by various optimizer passes like PRE, induction variable detection etc. and appropriate
+priorities are given to the candidates based on the number of references and live
+range (number of blocks). Advantages are that most of the code is platform neutral
```
Comment (dsouzai) [inline doc/compiler/optimizer/GlobalOptimizationsSummary.md:105]:
```suggestion
range (number of blocks). The advantages are that most of the code is platform neutral
```

File: doc/compiler/optimizer/GlobalOptimizationsSummary.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+and it is relatively cheap in compile time in comparison to (expected cost of) a
```
Comment (dsouzai) [inline doc/compiler/optimizer/GlobalOptimizationsSummary.md:106]:
```suggestion
and it is relatively cheap in compile time in comparison to (the expected cost of) a
```

File: doc/compiler/optimizer/GlobalOptimizationsSummary.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+graph-coloring allocator.
+
+## Live Variables for GC (local and global)
+Live Variables for GC attempts to minimize the number of locals that need to be zero
+initialized upon method entry. Basic idea is to perform liveness analysis and find
```
Comment (dsouzai) [inline doc/compiler/optimizer/GlobalOptimizationsSummary.md:111]:
```suggestion
initialized upon method entry. The basic idea is to perform liveness analysis and find
```

File: doc/compiler/optimizer/GlobalOptimizationsSummary.md [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+out if any locals are live at any [GC point](https://github.com/eclipse/omr/blob/81b79405da6c7c960e611a8b2b12fd5861543330/compiler/il/OMRNode.cpp#L2298)
+that is not dominated by (user defined) stores on all control flow paths. If a local
+is uninitialized at a GC point, explicit zero initialization must be inserted at the method entry.
+Note that this situation is fairly rare (only occurs in some uncommon cases involving jsr).
```
Comment (dsouzai) [inline doc/compiler/optimizer/GlobalOptimizationsSummary.md:115]: "Involving JSR": do you mean JSR292? JSR by itself just refers to Java Specification Request.

File: doc/compiler/optimizer/LocalOptimizationsSummary.md [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+# Local Optimizations
+
+Local optimizations operate on extended basic blocks at a time.
+
+## Tree Simplification
+Tree Simplification performs simple IL tree transformations like constant
+folding, strength reduction, canonicalizing expressions (e.g. an `add` with a
+constant child will have the constant as the second child), branch/switch
+simplification etc. Float/double simplification support needs to be improved
+(call C libraries to compute the correct answer as per the Java spec).
```
Comment (dsouzai) [inline doc/compiler/optimizer/LocalOptimizationsSummary.md:32]: When you say support needs to be improved, are you calling out a shortcoming of the current state? If so, it probably doesn't make sense to have it in a documentation setting, unless you can reword it to describe what the current simplification does and that it has opportunity for further refinement.

File: doc/compiler/optimizer/LocalOptimizationsSummary.md [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+## Local Value Propagation
+Local [Value Propagation](https://github.com/eclipse/omr/blob/master/doc/compiler/optimizer/ValuePropagation.md)
+performs constant, type and relational propagation within a block.
+It propagates constants and types based on assignments and performs removal
+of subsumed checks and compares. It also performs devirtualization
+of calls, checkcast and array store check elimination. It does not do value
+numbering, instead it works on the assumption that every expression has a
+unique value number (improved effectiveness if expressions having same value
+are commoned by a prior local CSE pass).
```
Comment (dsouzai) [inline doc/compiler/optimizer/LocalOptimizationsSummary.md:49]:
```suggestion
unique value number; its effectiveness is improved if expressions having same value
are commoned by a prior local CSE pass.
```

File: doc/compiler/optimizer/LocalOptimizationsSummary.md [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+there are two stores to the same memory without any intervening use in the
+same block).
+
+## Local Reordering of Expressions
+Local Reordering of Expressions reduces live ranges (register pressure) by
+moving evaluations around within a block.
+
+## Array Length Simplification
+In OpenJ9, array lengths are stored in number of bytes. So bound checks require
+a shift right (depending on the array stride) on the array length so that
```
Comment (dsouzai) [inline doc/compiler/optimizer/LocalOptimizationsSummary.md:62]:
```suggestion
a shift right (depending on the array stride) on the array length so that the
```

File: doc/compiler/optimizer/LocalOptimizationsSummary.md [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+compare can be done with array index. This optimization attempts to eliminate
```
Comment (dsouzai) [inline doc/compiler/optimizer/LocalOptimizationsSummary.md:63]:
```suggestion
compare can be done with the array index. This optimization attempts to eliminate
```

File: doc/compiler/optimizer/LocalOptimizationsSummary.md [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+`shr` wherever possible (instead comparing the index left shifted based on
```
Comment (dsouzai) [inline doc/compiler/optimizer/LocalOptimizationsSummary.md:64]:
```suggestion
`shr` wherever possible (instead comparing the index left shifted based on the
```

File: doc/compiler/optimizer/LocalOptimizationsSummary.md [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+that it is legal to do so. Note that shifting left may cause wrap-arounds (if
+the value is close to MAXINT for example) and so this optimization is done
+for constant indices or indices marked as safe to left shift by value propagation.
+Shifting left is better as it is already required in most cases for the address
+computation for the array access.
+
+## Compaction of Null Explicit Null Checks
+Compaction of Null Explicit Null Checks minimize the cases where an explicit null
+check needs to be performed (null checks are implicit, i.e. no code for doing null
+check, instead we use hardware trap on IA32). Sometimes (say as a result of dead
```
Comment (dsouzai) [inline doc/compiler/optimizer/LocalOptimizationsSummary.md:75]:
```suggestion
check needs to be performed (null checks are implicit, i.e. by default on IA32, the JIT does not
emit code to perform a null check, instead we use a hardware trap). Sometimes (e.g. as a result of dead
```

Comment (dsouzai) [reply doc/compiler/optimizer/LocalOptimizationsSummary.md:75]: Do we have implicit null checks only on IA32?

File: doc/compiler/optimizer/LocalOptimizationsSummary.md [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+nullness implicitly); this pass attempts to find another expression in the same
+block that can be used to perform the null check implicitly again.
+
+## Dead Trees Elimination
+Expressions that are evaluated and not used for a few instructions can actually be
+evaluated exactly where required (subject to the constraint that their values should
+be the same if evaluated later). This optimization attempts to simply delay evaluation
+of expressions and has the effect of reducing register pressure by shortening
+live ranges. It has the effect of removing some IL trees that are simply anchors
+for (already) evaluated expressions and making the trees more compact.
```
Comment (dsouzai) [inline doc/compiler/optimizer/LocalOptimizationsSummary.md:87]:
```suggestion
for (already) evaluated expressions thus making the trees more compact.
```

File: doc/compiler/optimizer/LocalOptimizationsSummary.md [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+
+## Rematerialization
+Rematerialization is performed as a late pass before code generation is to be done.
```
Comment (dsouzai) [inline doc/compiler/optimizer/LocalOptimizationsSummary.md:90]:
```suggestion
Rematerialization is performed as a late pass before code generation.
```

File: doc/compiler/optimizer/LocalOptimizationsSummary.md [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+Basically queries the underlying code generator (IA32 etc.) for the max number of
```
Comment (dsouzai) [inline doc/compiler/optimizer/LocalOptimizationsSummary.md:91]:
```suggestion
The optimization queries the underlying code generator (IA32 etc.) for the max number of
```

File: doc/compiler/optimizer/LocalOptimizationsSummary.md [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+registers available for assignment simultaneously, and then tries to reverse
+commoning in regions of code within a block where number of live expressions could
```
Comment (dsouzai) [inline doc/compiler/optimizer/LocalOptimizationsSummary.md:93]:
```suggestion
commoning in regions of code within a block where the number of live expressions could
```

File: doc/compiler/optimizer/LocalOptimizationsSummary.md [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+cause spills. Only certain kinds of expressions are considered for rematerialization,
+e.g. address calculations (can often use more complex addressing modes on a chip),
```
Comment (dsouzai) [inline doc/compiler/optimizer/LocalOptimizationsSummary.md:95]:
```suggestion
e.g. address calculations (as chips can support more complex addressing modes),
```

File: doc/compiler/optimizer/LoopOptimizationSummary.md [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+with a loop invariant (pre-header) block. It places the loop test at the end
+of the trees for the loop, so that the loop back-edge is almost always a
+backwards branch.
+
+## Loop Versioning and Loop Specialization
+During Loop Versioning, loop invariant null checks, bound checks, div checks,
+checkcasts and inline guards (where the `this` does not change in the loop)
+are eliminated from loops by placing equivalent tests outside the loop. It
+also eliminates bound checks where the range of values of the index inside
+the loop is discovered (need to be a compile time constant) and checks are
```
Comment (dsouzai) [inline doc/compiler/optimizer/LoopOptimizationSummary.md:36]:
```suggestion
the loop is discovered (if it is a compile time constant) and checks are
```

File: doc/compiler/optimizer/LoopOptimizationSummary.md [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+## Loop Invariant Code Motion
+Done as part of [Partial Redundancy Elimination (PRE)](optimizer/GlobalOptimizationsSummary.md)
+
+## Loop Inversion
+Loop Inversion converts a loop where the induction variable is counting up
+from zero into one where it counts down to zero. Note that this can only be
+done if the inversion of the loop does not affect program semantics inside
+the loop (order of exceptions thrown etc). The benefit of doing this is the
+fact that there are, in general, instructions that can perform compare/branch
+against zero in an efficient manner and special counter machine register
```
Comment (dsouzai) [inline doc/compiler/optimizer/LoopOptimizationSummary.md:74]:
```suggestion
against zero in an efficient manner and special counter machine registers
```

File: doc/compiler/optimizer/LoopOptimizationSummary.md [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+(for e.g. On PowerPC) can be used.
+
+## Loop Striding
+Loop Striding creates derived induction variables (e.g. the address calculation
+for array accesses `&a[i]`) and increment/decrement the derived induction
+variables by the appropriate stride through every iteration of the loop.
+This can eliminate some complex array address calculations inside the loop,
+instead replacing by a load (reg or memory). Note that we also create derived
```
Comment (dsouzai) [inline doc/compiler/optimizer/LoopOptimizationSummary.md:82]:
```suggestion
This can eliminate some complex array address calculations inside the loop,
replaced instead by a load (reg or memory). Note that we also create derived
```

File: doc/compiler/optimizer/LoopOptimizationSummary.md [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+
+## Async Check (yield point) Placement/Removal
+The algorithm for minimizing the async checks in a given loop attempts to
+achieve its goal by adopting two distinct methods:
+
+- Check if the loop is known to be short running (either based on induction
+variable information obtained from [Value Propagation](https://github.com/eclipse/omr/blob/master/doc/compiler/optimizer/ValuePropagation.md))
+or if the special versioning test for short running loops (mentioned earlier)
+is found. In either case, since the loop is provably not long running and does
+not require an async check at all.
```
Comment (dsouzai) [inline doc/compiler/optimizer/LoopOptimizationSummary.md:109]:
```suggestion
is found. In either case, since the loop is provably not long running it does
not require an async check at all.
```

File: doc/compiler/optimizer/LoopOptimizationSummary.md [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+- If async checks cannot be skipped completely for the loop, attempt to place
+the sync checks at points within the loop such that there is minimal redundancy
```
Comment (dsouzai) [inline doc/compiler/optimizer/LoopOptimizationSummary.md:111]:
```suggestion
the async checks at points within the loop such that there is minimal redundancy
```

**Review by dsouzai — COMMENTED** [review]
Thanks for writing this @a7ehuo; it was a very informative read. Mostly minor changes requested.

**Comment by a7ehuo** [comment]
>Thanks for writing this @a7ehuo; 

Just to clarify, I didn't write these documents. The original document is written by @vijaysun-omr. I just broke it down into the four individual documents and edited them a bit.

**Comment by dsouzai** [comment]
Heh, well thanks for opening the PR notwithstanding :).

File: doc/compiler/optimizer/LoopOptimizationSummary.md [hunk 23]
```diff
@@ -0,0 +1,114 @@
...
+During Loop Versioning, loop invariant null checks, bound checks, div checks,
+checkcasts and inline guards (where the `this` does not change in the loop)
+are eliminated from loops by placing equivalent tests outside the loop. It
+also eliminates bound checks where the range of values of the index inside
+the loop is discovered (need to be a compile time constant) and checks are
+inserted outside the loop so that no bound checks fail inside the loop.
+
+This analysis uses induction variable information found by [Value Propagation](https://github.com/eclipse/omr/blob/master/doc/compiler/optimizer/ValuePropagation.md).
+Another versioning test that is emitted ensures the loop will not run for a
+very long period of time (in which case the async check inside the loop can
```
Comment (hzongaro) [inline doc/compiler/optimizer/LoopOptimizationSummary.md:41]: Is there any place that describes what an async check?

File: doc/compiler/optimizer/GlobalOptimizationsSummary.md [hunk 24]
```diff
@@ -0,0 +1,138 @@
...
+
+Local Live Variables for GC does not perform a full blown liveness analysis; instead
+it performs a forward walk through the CFG till the first GC point is seen and collects
+all the locals that have been stored till that point and marks the other locals as
+requiring zero initialization.
+
+Global Live Variables for GC collects zero initialization information as explained above;
+in addition it also populates each basic block with information about which locals
+are live on entry to the block. This is then used by the code generator to only
+(selectively) switch on bits in the GC stack maps for those reference locals that
```
[the remark this answers is not in the document]
Comment (vijaysun-omr) [reply doc/compiler/optimizer/GlobalOptimizationsSummary.md:127]: There is some historical documentation that @0xdaryl had found related to GC stack maps. Maybe he can comment on whether there was some plan to move to the open in the future.

[the remark this answers is not in the document]
Comment (0xdaryl) [reply doc/compiler/optimizer/GlobalOptimizationsSummary.md:127]: I believe the material Vijay mentions could be used as a starting point for a GC map overview but more will need to be added to it to make it a useful contribution.  This is on our rather long list of concepts we wish to document as part of the tech debt initiative, but no document exists just yet to link to in this document..

File: doc/compiler/optimizer/ControlFlowOptimizationsSummary.md [hunk 25]
```diff
@@ -0,0 +1,65 @@
...
+[1] https://www.gnu.org/software/classpath/license.html
+[2] http://openjdk.java.net/legal/assembly-exception.html
+
+SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+-->
+
+# Control Flow Optimizations
+
+## Block Reordering
+Block Reordering moves blocks around based on making the ‘hot’ path (based on
```
[the remark this answers is not in the document]
Comment (vijaysun-omr) [reply doc/compiler/optimizer/ControlFlowOptimizationsSummary.md:26]: Might be better to name this section `Basic block ordering` to fit in with the actual name of the optimization in the strategy, that is `basicBlockOrdering`

File: doc/compiler/optimizer/ControlFlowOptimizationsSummary.md [hunk 26]
```diff
@@ -0,0 +1,65 @@
...
+profiling and/or static criteria like loop nesting depth, expected code synergy)
+the fall through.
+
+## Block Hoisting
+Block Hoisting duplicates and places the code in a block into its predecessors
+that only have the block as a unique successor, if this would enable better/more
+commoning or other optimizations to occur.
```
[the remark this answers is not in the document]
Comment (vijaysun-omr) [reply doc/compiler/optimizer/ControlFlowOptimizationsSummary.md:33]: I would add one extra sentence : `One example of when this optimization may be effective is with respect to local dead store elimination when the hoisted block ends in a return, since might allow the removal of trailing stores to local variables.`

File: doc/compiler/optimizer/ControlFlowOptimizationsSummary.md [hunk 27]
```diff
@@ -0,0 +1,65 @@
...
+
+## Block Splitting
+Block Splitting is similar to virtual guard tail splitting except that profiled
```
[the remark this answers is not in the document]
Comment (vijaysun-omr) [reply doc/compiler/optimizer/ControlFlowOptimizationsSummary.md:36]: We may want to consider adding : `The basic idea of block (and virtual guard tail) splitter are the same : namely, do the transformation referred to as "tail duplication" in compiler literature, that essentially involves duplicating code after control flow merge points to avoid the merge in control flow. This transformation should result in more opportunity to specialize and optimize the duplicated code for the hottest paths leading up to the control flow merge.`

File: doc/compiler/optimizer/GlobalOptimizationsSummary.md [hunk 28]
```diff
@@ -0,0 +1,138 @@
...
+completely independently of the others. Note that this can only be done if the allocation
+is not used as an object (e.g. passed out into a call which dereferences the object).
+
+- Contiguous allocations mean that the object is allocated on the stack but the fields
+are all part of the object on the stack. References to the fields still have to use
+the underlying object (except that it's on the stack instead of the heap).
+
+An early pass that splits (duplicates) one allocation into multiple allocations along
+different control flow paths (allocation splitting) is used to maximize the opportunities
+for escape analysis.
```
[the remark this answers is not in the document]
Comment (vijaysun-omr) [reply doc/compiler/optimizer/GlobalOptimizationsSummary.md:62]: I agree with @hzongaro There probably is an escape analysis talk and/or presentation from him that we could link to if it is in the open. 

Other points that we probably should mention : `There is a notion of "heapifying" a stack allocated object in cases even when an escape point is detected for the object but the escape point is a cold block. In such cases it is worth keeping the object on stack on the hot path and only if the cold path is taken, then the object is allocated on the heap and the contents of the object on the stack are copied into the newly allocated heap object. As part of heapification, all references from autos/parms to the stack allocated object are updated to point to the heapified object instead. This means that it is safe to execute any code where the object escapes from that point on.`

[the remark this answers is not in the document]
Comment (0xdaryl) [reply doc/compiler/optimizer/GlobalOptimizationsSummary.md:62]: EA overview -> https://youtu.be/JMpqwLVwmk8

[the remark this answers is not in the document]
Comment (vijaysun-omr) [reply doc/compiler/optimizer/GlobalOptimizationsSummary.md:62]: Other potential additions:

`Our escape analysis implementation also has a notion of de-memoization/re-memoization. This pertains to some special classes in the JCL: Integer, Long, Short, Byte etc. that memoize some instances of those classes in a static array that they maintain (only for some ranges of values). These classes would use these arrays to return an already allocated/memoized object rather than allocate a new object every time (if possible) when `valueOf` is called. These arrays therefore also become escape points for objects that are allocated and stored into them. De-memoization attempts to get around this by converting valueOf calls to new allocations before escape analysis is attempted on the assumption that allocating such objects on the stack would be even more preferable to memoizing them in the arrays. If stack allocation was not possible due to some reason, the de-memoized allocations are re-memoized, leaving the code as it was originally and an attempt is also made to inline the re-memoized valueOf since the inlining of these calls is suppressed in the inliner to give de-memoization a chance later.`

[the remark this answers is not in the document]
Comment (vijaysun-omr) [reply doc/compiler/optimizer/GlobalOptimizationsSummary.md:62]: Finally, on the Power platform (maybe AArch64 should use it too if it has a weak memory model, attn : @0xdaryl ) there is a "flow sensitive escape analysis" pass that gets run. 

`Flow sensitive escape analysis tries to find places in the code where an object "has not escaped yet" rather than the notion of "does the object escape anywhere" that is the hallmark of the flow insensitive escape analysis that runs on non-Power platforms. The reason for doing this flow sensitive analysis is because we want to optimize away the lwsync instructions that must be emitted after allocations to ensure that other threads see a properly initialized object. Flow sensitive escape analysis attempts to move such lysyncs together (i.e. combine) as much as possible. This includes trying to move an lwsync for an object to another lwsync (either for the same object (after a constructor if the object has final fields) or a different object), a monexit or volatile store (that do lwsyncs for different reasons) but all only on the basis that the object has not escaped yet.`

File: doc/compiler/optimizer/GlobalOptimizationsSummary.md [hunk 29]
```diff
@@ -0,0 +1,138 @@
...
+
+## Explicit New Initialization
+Explicit New Initialization attempts to skip zero initialization of objects where possible.
+Basically it scans forward in the basic block from the point of the allocation and
+tries to find explicit definitions of fields (say in constructors) that occur before
+any use of the field. In these cases the zero initialization of fields (required according
+to Java spec, so no field is uninitialized) can be skipped. Note that for reference
+fields the presence of a GC point before any user-specified definition means that
```
[the remark this answers is not in the document]
Comment (vijaysun-omr) [reply doc/compiler/optimizer/GlobalOptimizationsSummary.md:70]: The notion of what qualifies as a GC point in the JIT typically depends on the query `node->canGCandReturn()`

File: doc/compiler/optimizer/GlobalOptimizationsSummary.md [hunk 30]
```diff
@@ -0,0 +1,138 @@
...
+on some object followed by another syncenter/syncexit on the same object). In this
+case the extent of code across which the lock is held is increased and one syncexit/syncenter
+can be removed. Deadlock avoidance is ensured by examining the code where the lock
+would be held after coarsening and making sure there are no calls or lock/unlock
+operations in that region of code. Sometimes additional syncenters/syncexits may be
+added on ‘cold’ paths to avoid some syncenters/syncexits on ‘hot’ paths.
+
+## Global Register Allocator (GRA)
+GRA is a tactical solution for achieving the effect of global register assignment,
+even though the underlying code generators have local register allocators. Basic idea
```
[the remark this answers is not in the document]
Comment (vijaysun-omr) [reply doc/compiler/optimizer/GlobalOptimizationsSummary.md:97]: This tactical solution has survived now for over 20 years, i.e. I'd suggest it has earned the right to no longer be referred to as tactical :) @0xdaryl

File: doc/compiler/optimizer/GlobalOptimizationsSummary.md (hunk 3 above)
Comment (vijaysun-omr) [reply doc/compiler/optimizer/GlobalOptimizationsSummary.md:105]: Suggestion : `most of the code is platform neutral and is` -> `most of the code is platform neutral apart from a few simple evaluators that must be added to a new platform code generator, thus drastically reducing the development cost of adding a powerful global register allocator. It is also `

File: doc/compiler/optimizer/GlobalOptimizationsSummary.md (hunk 6 above)
Comment (vijaysun-omr) [reply doc/compiler/optimizer/GlobalOptimizationsSummary.md:115]: No, a long long time ago in a galaxy far far away, there used to be a bytecode known as `jsr` (jump to sub routine) that was used to implement throw/catch/finally when Java was first released. Because this code essentially caused some abnormal control flow, it could make it appear as though a variable was uninitialized on some code path. The `jsr` bytecode was removed a long time back. 

I believe the main situation in which such zero initializations may be needed in today's JIT compiler may be around some optimization creating such an appearance of uninitialized variables. 

I wonder if we should add the word "reference"  to "local" since the GC cares only about reference locals.

File: doc/compiler/optimizer/LocalOptimizationsSummary.md [hunk 31]
```diff
@@ -0,0 +1,103 @@
...
+
+[1] https://www.gnu.org/software/classpath/license.html
+[2] http://openjdk.java.net/legal/assembly-exception.html
+
+SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+-->
+
+# Local Optimizations
+
+Local optimizations operate on extended basic blocks at a time.
```
[the remark this answers is not in the document]
Comment (vijaysun-omr) [reply doc/compiler/optimizer/LocalOptimizationsSummary.md:25]: Yes, they operate on a single extended basic block at a time.

File: doc/compiler/optimizer/LocalOptimizationsSummary.md (hunk 9 above)
Comment (vijaysun-omr) [reply doc/compiler/optimizer/LocalOptimizationsSummary.md:62]: This optimization no longer is needed since arraylengths were changed in OpenJ9 to be stored as "number of elements" rather than "number of bytes" 20 years ago :)

Comment (vijaysun-omr) [reply doc/compiler/optimizer/LocalOptimizationsSummary.md:62]: To be 100% clear, I am saying this optimization does not exist for the above reason anymore and so it also should not be documented.

File: doc/compiler/optimizer/LocalOptimizationsSummary.md (hunk 12 above)
Comment (vijaysun-omr) [reply doc/compiler/optimizer/LocalOptimizationsSummary.md:75]: No, we have them on X86 (32 and 64 bitness). IA32 was just the quaint way of referring to the platform when "64-bit was new".

File: doc/compiler/optimizer/LoopOptimizationSummary.md (hunk 23 above)
Comment (vijaysun-omr) [reply doc/compiler/optimizer/LoopOptimizationSummary.md:41]: Not to my knowledge. Async checks get added by the VM at backward branches in the bytecodes on the assumption that one cannot have a loop without a backwards branch. The async check is used for many purposes all of which are some variant of the need to periodically have all application threads yield in order for some sort of action to be taken, with the prime examples being a) GC and b) walk the stack to sample where the thread is executing for driving compilation control. A better definition may be possible, but that hopefully gives the gist.

File: doc/compiler/optimizer/LoopOptimizationSummary.md [hunk 32]
```diff
@@ -0,0 +1,114 @@
...
+## Loop Unrolling and Peeling
+A majority of loops can be unrolled or peeled with or without the loop driving
+tests done after each iteration (loop) body. Usually a peel is inserted before
+the unrolled loop and the residual iterations to be done (at most u-1 residual
+iterations where u is the unroll factor) are also done using the peeled code.
+Peeling aids [Partial Redundancy Elimination (PRE)](optimizer/GlobalOptimizationsSummary.md)
+as code does not have to be moved; instead commoning of dominated expressions can be done
+(which is far easier than code motion because of problems introduced by exception checks
+in Java).
+
```
Comment (vijaysun-omr) [inline doc/compiler/optimizer/LoopOptimizationSummary.md:59]: We had a plan to implement loop peeling but we never actually did it. So the documentation should not be included for it.

File: doc/compiler/optimizer/GlobalOptimizationsSummary.md (hunk 29 above)
[the remark this answers is not in the document]
Comment (a7ehuo) [reply doc/compiler/optimizer/GlobalOptimizationsSummary.md:70]: There is a hyperlink link to node->canGCandReturn() in a later section `Live Variables for GC (local and global)`. I'll move that link to the first time when GC point is mentioned. 

```
## Live Variables for GC (local and global)
Live Variables for GC attempts to minimize the number of locals that need to be zero
initialized upon method entry. The basic idea is to perform liveness analysis and find
out if any locals are live at any [GC point](https://github.com/eclipse/omr/blob/81b79405da6c7c960e611a8b2b12fd5861543330/compiler/il/OMRNode.cpp#L2298)
```

**Comment by a7ehuo** [comment]
@vijaysun-omr @dsouzai @hzongaro 3becd9a addresses all the above comments except for the following two comments. I think they can always be updated once we have an initial version of these documents. It's ready for [another review](https://github.com/eclipse-openj9/openj9/compare/f3755b692379cca72a695e96396d8e03362d09d5..3becd9a659bf3949491c6d694b145d1c75c177a6). Thank! 

1. Add a small description for some pre and post escape analysis passes that were added for handling HCR better https://github.com/eclipse-openj9/openj9/pull/14999#discussion_r869253165
2. Add a link to GC stack map document: https://github.com/eclipse-openj9/openj9/pull/14999#discussion_r868043884

File: doc/compiler/optimizer/GlobalOptimizationsSummary.md [hunk 33]
```diff
@@ -0,0 +1,214 @@
...
+trivial control flow between the `new` and the `<init>` call.
+
+On PowerPC platform, there is a "Flow Sensitive Escape Analysis" pass.
+Flow Sensitive Escape Analysis tries to find places in the code where an object
+"has not escaped yet" rather than the notion of "does the object escape anywhere" that is
+the hallmark of the Flow Insensitive Escape Analysis that runs on non-Power platforms.
+The reason for doing this Flow Sensitive Escape Analysis is because we want to optimize away the
+`lwsync` instructions that must be emitted after allocations to ensure that other threads
+see a properly initialized object. Flow Sensitive Escape Analysis attempts to move such
+`lysyncs` together (i.e. combine) as much as possible. This includes trying to move an `lwsync`
```
Comment (vijaysun-omr) [inline doc/compiler/optimizer/GlobalOptimizationsSummary.md:124]: typo : `lysync` -> `lwsync`

**Review by vijaysun-omr — APPROVED** [review]

**Comment by vijaysun-omr** [comment]
Apart from the minor typo I mentioned in my review this morning, I am fine with this version. It can be added to with other optimizations later as a separate work item. Thanks @a7ehuo

**Review by hzongaro — APPROVED** [review]
Looks good - Thanks!

**Review by dsouzai — APPROVED** [review]

**Comment by dsouzai** [comment]
Merging as all reviewers have approved.
